codec-http2/src/main/java/io/netty/handler/codec/http2/HpackEncoder.java

```diff
@@ -380,142 +402,160 @@ private HeaderEntry getEntryInsensitive(CharSequence name, CharSequence value) {
+        latest.after = e;
+        latest = e;
     }
 
     /**
-     * Add the header field to the dynamic table. Entries are evicted from the dynamic table until
-     * the size of the table and the new header field is less than the table's maxHeaderTableSize. If the size
-     * of the new entry is larger than the table's maxHeaderTableSize, the dynamic table will be cleared.
+     * Remove the oldest header field from the dynamic table.
      */
-    private void add(CharSequence name, CharSequence value, long headerSize) {
-        // Clear the table if the header field size is larger than the maxHeaderTableSize.
-        if (headerSize > maxHeaderTableSize) {
-            clear();
-            return;
+    private void remove() {
+        NameValueEntry eldest = head.after;
+        removeNameValueEntry(eldest);
+        removeNameEntryMatchingCounter(eldest.name, eldest.counter);
+        head.after = eldest.after;
+        eldest.remove();
+        size -= eldest.size();
+        if (isEmpty()) {
+            latest = head;
         }
+    }
 
-        // Evict oldest entries until we have enough maxHeaderTableSize.
-        while (maxHeaderTableSize - size < headerSize) {
-            remove();
-        }
+    private boolean isEmpty() {
+        return size == 0;
+    }
 
-        int h = AsciiString.hashCode(name);
-        int i = index(h);
-        HeaderEntry old = headerFields[i];
-        HeaderEntry e = new HeaderEntry(h, name, value, head.before.index - 1, old);
-        headerFields[i] = e;
-        e.addBefore(head);
-        size += headerSize;
+    private void removeNameValueEntry(NameValueEntry eldest) {
+        int bucket = bucket(eldest.hash);
+        NameValueEntry e = nameValueEntries[bucket];
+        if (e == eldest) {
+            nameValueEntries[bucket] = eldest.next;
+        } else {
+            while (e.next != eldest) {
+                e = e.next;
+            }
+            e.next = eldest.next;
+        }
     }
 
-    /**
-     * Remove and return the oldest header field from the dynamic table.
-     */
-    private HpackHeaderField remove() {
-        if (size == 0) {
-            return null;
+    private void removeNameEntryMatchingCounter(CharSequence name, int counter) {
+        int hash = AsciiString.hashCode(name);
+        int bucket = bucket(hash);
+        NameEntry e = nameEntries[bucket];
+        if (e == null) {
+            return;
         }
-        HeaderEntry eldest = head.after;
-        int h = eldest.hash;
-        int i = index(h);
-        HeaderEntry prev = headerFields[i];
-        HeaderEntry e = prev;
-        while (e != null) {
-            HeaderEntry next = e.next;
-            if (e == eldest) {
-                if (prev == eldest) {
-                    headerFields[i] = next;
-                } else {
-                    prev.next = next;
+        if (counter == e.counter) {
+            nameEntries[bucket] = e.next;
+            e.remove();
```

Comment:
Nit: we could have `remove()` return `next` and so collapse.

Reply:
It will make the code shorter, but a bit more cryptic:

1. `remove` in common Java API returns the item that was removed, here we will be returning the next item.
2. This method does not remove anything, it just clears pointers to eliminate possible old -> young generation reference. Renamed it to `unlink`.